src/board/system76/common/flash/main.c

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (C) 2020 Evan Lojewski
+ *
+ * SPDX-License-Identifier: GPL-3.0-or-later
+ */
+
+#include <stdint.h>
+#include <board/flash.h>
+
+// EC indirect flash access
+volatile uint8_t __xdata __at(0x103B) ECINDAR0;
+volatile uint8_t __xdata __at(0x103C) ECINDAR1;
+volatile uint8_t __xdata __at(0x103D) ECINDAR2;
+volatile uint8_t __xdata __at(0x103E) ECINDAR3;
+volatile uint8_t __xdata __at(0x103F) ECINDDR;
+
+#define SPI_DEVICE                  (0x70)
+#define SPI_FOLLOW_MODE             (0x0F)
+#define SPI_CHIP_SELECT             (0xFD)
+#define SPI_CHIP_DESELECT           (0xFE)
+
+#define SPI_READ_STATUS_COMMAND     (0x05)
+#define SPI_READ_COMMAND            (0x0B)
+#define SPI_WRITE_COMMAND           (0x02)
+#define SPI_WRITE_ENABLE_COMMAND    (0x06)
+#define SPI_READ_STATUS_COMMAND     (0x05)
+
+#define SPI_ERASE_SECTOR_COMMAND    (0xD7)
+
+#define SPI_STATUS_WIP              (0x01)
+
+void flash_write_enable(void);
+void flash_wait(void);
+void flash_enter_follow_mode(void);
+void flash_exit_follow_mode(void);
+
+// Main flash API entry point. Note: marked as __reentrant so that the main application __data variables are not stomped on.
+void flash_entry(uint32_t addr, uint8_t buffer[], uint8_t size, uint8_t command) __reentrant __critical {
+    if(FLASH_COMMAND_READ == command) {
```

Comment:
fmt: Missing space between control statement and opening brace
nit: Yoda condition

Reply:
fmt: fixed.
Per to the Yoda condition: This is done on purpose. The primary reason is that if the == is accidentally typed as an =, you'll get this error in sdcc:
src/board/system76/common/flash/main.c:39: error 33: Attempt to assign value to a constant variable (=)

If it's the other way around, no error is attributed to that line. That said, sdcc does throw other errors later on:
src/board/system76/common/flash/main.c:53: error 110: conditional flow changed by optimizer: so said EVELYN the modified DOG
src/board/system76/common/flash/main.c:85: error 110: conditional flow changed by optimizer: so said EVELYN the modified DOG
src/board/system76/common/flash/main.c:40: error 126: unreachable code

Basically, on good compilers (clang), it'll throw an error about it and specifically state that if it's intended, you need to wrap it in an additional set of ()'s. On bad compilers, no error will be thrown. On sdcc, no error is thrown but it does cause errors later on to be thrown.

It's part of the coding standards that I usually follow for that specific reason. I'd prefer to leave it as is, but if you want conditionals to be the other way, I can change them.

Reply:
The convention used throughout the code is to have constants on the right hand side. We don't enforce any style (yet), but I would prefer to keep it consistent with existing code.